### ISSUE TITLE
[ISSUE] Provider issue: Authentication with azurerm  Version > 2.78

### ISSUE THREAD

**Issue by jkritzen** [databrickslabs/terraform-provider-databricks#1137]
Hi there,

Databricks provider breaks with azurerm version > 2.78 if Databricks Workspace is already deployed. 

### Configuration
```hcl
terraform {
  required_providers {
    azurerm = {
      source  = "hashicorp/azurerm"
      version = "2.97.0"
    }
    databricks = {
      source  = "databrickslabs/databricks"
      version = "0.5.0"
    }
}

# use Provider DataBricks for Cluster Setup
provider "databricks" {
  host                        = azurerm_databricks_workspace.ws.workspace_url
  azure_workspace_resource_id = azurerm_databricks_workspace.ws.id
  azure_client_id             = azuread_application.app_adb.application_id
  azure_client_secret         = azuread_application_password.app_adb.value
  azure_tenant_id             = data.azurerm_client_config.current.tenant_id
  alias                       = "ws"
  azure_use_msi               = true
}
resource "databricks_cluster" "single_cluster" {
  depends_on = [
    azurerm_role_assignment.contributor_adb,
    resource.azurerm_databricks_workspace.ws
  ]
  provider                = databricks.ws
  cluster_name            = "${local.databricks_name}-SINGLE"
  spark_version           = "9.1.x-scala2.12" #data.databricks_spark_version.latest_lts.id
  node_type_id            = "Standard_F4s"    # data.databricks_node_type.smallest.id
  autotermination_minutes = 120
  num_workers             = 0
  spark_conf = {
    # Single-node
    "spark.databricks.cluster.profile" : "singleNode"
    "spark.master" : "local[*]"
    "spark.databricks.delta.retentionDurationCheck.enabled" : "false"
    "spark.executor.cores" : "1"
    "spark.executor.memory" : "1g"
  }
  custom_tags = {
    "ResourceClass" = "SingleNode"
  }
}
```

### Expected Behavior
Should be not result in Authentication Error. If using provider **registry.terraform.io/hashicorp/azurerm v2.78.0** everything works like expected.

### Actual Behavior
│ Error: cannot read cluster: cannot configure azure-client-secret auth: cannot get workspace: please set `azure_workspace_resource_id` provider argument. Environment variables used: ARM_CLIENT_SECRET, ARM_CLIENT_ID, ARM_TENANT_ID. Please check https://registry.terraform.io/providers/databrickslabs/databricks/latest/docs#authentication for details
│ 
│   with databricks_cluster.single_cluster,
│   on databricks.tf line 181, in resource "databricks_cluster" "single_cluster":
│  181: resource "databricks_cluster" "single_cluster" {
│ 
╵

### Steps to Reproduce
Please list the steps required to reproduce the issue, for example:
1. `terraform plan` with existing Databricks Workspace and update of azurerm to actual Version.

### Terraform and provider versions
Terraform: Terraform v1.1.6 on linux_amd64

Providers:
- Installed databrickslabs/databricks v0.5.0 (signed by a HashiCorp partner, key ID 905AA25F2E92C2D5)
- Installed hashicorp/azuread v2.18.0 (signed by HashiCorp)
- Installed hashicorp/time v0.7.2 (signed by HashiCorp)
- Installed hashicorp/azurerm v2.97.0 (signed by HashiCorp)

### Important Factoids
Similar Issue: https://github.com/databrickslabs/terraform-provider-databricks/issues/1045

Downgrade to azurerm 2.78 works, but i need the actual version.

**Comment by nfx** [comment]
Also very similar to #781

**Comment by jkritzen** [comment]
Yes, similar issue. I am not sure if the root cause is within the azurerm provider or the databricks provider.

**Comment by nfx** [comment]
I've opened a thread in azurerm slack - https://terraform-azure.slack.com/archives/CB9RVKPDL/p1645610498112489, let's see the outcome.

@jkritzen could you do a favor and do some azurerm [version bisecting](https://en.wikipedia.org/wiki/Bisection_(software_engineering)) and telling when it stops working?.. e.g. 2.97-2.78 - 19 versions, so first try pinning to 2.88, if it works - pin to 2.94, if it doesn't - pin to 2.82, and so on ... in 3-4 attempts you'll get the exact `azurerm` version that breaks things.

**Comment by zaher-mh** [comment]
Hello ! We have the exact same issue on a project with TF and Databricks on Azure

**Comment by jkritzen** [comment]
Hi, 
did not have the time yet for testing. It could be that i can test tomorrow.

Kind Regards

**Comment by nfx** [comment]
@jkritzen Please do. For now azurerm team states that there weren't any changes that could break it and the problem is in the design of Terraform Core itself at the moment.

**Comment by nfx** [comment]
@jkritzen any update on the exact version?

**Comment by jkritzen** [comment]
Hi,
i had a new setup yesterday (additional to the existing environment) and it seems that this issue is when you change the provider versions of an existing deployment.

Currently the new deployment works with the following versions:

- 2022-03-03T16:02:58.4979499Z     ├── provider[registry.terraform.io/hashicorp/azurerm] 2.97.0
- 2022-03-03T16:02:58.4980235Z     └── provider[registry.terraform.io/databrickslabs/databricks] 0.5.0

**Comment by nfx** [comment]
so, the issue is fixed in 2.97?

**Comment by trapacampo** [comment]
Hello! I have the same issue when trying to get the spark version with data.databricks_spark_version.

**Comment by trapacampo** [comment]
@tbugfinder yes, it's possible. I created a single module to create the Databricks workspace and the Databricks cluster using the provider.
You have to use depends_on = [azurerm_databricks_workspace.example] in all databricks resources/data that use the databricks providers.
Explained here: https://registry.terraform.io/providers/databrickslabs/databricks/latest/docs#data-resources-and-authentication-is-not-configured-errors

Hope it helps :)

**Comment by jkritzen** [comment]
@tbugfinder : This is from terraform not only a single step, as terraform calculated the dependend ressource individual which can done between the providers with the meta argument depends_on = []

The tricky one, is that creating Service Principals takes some time, maybe you need an additional wait timer.

**Comment by tbugfinder** [comment]
> This is from terraform not only a single step, as terraform calculated the dependend ressource individual which can done between the providers with the meta argument depends_on = []

Indeed I missed that, thx. I really got hesitant in using depends_on due to its side effects but will try it now.

**Comment by tbugfinder** [comment]
> Indeed I missed that, thx. I really got hesitant in using depends_on due to its side effects but will try it now.

as expected data providers fail with this kind of depens_on configuration as those get evaluated to "early" - as far as I understand.

**Comment by trapacampo** [comment]
> > Indeed I missed that, thx. I really got hesitant in using depends_on due to its side effects but will try it now.
> 
> as expected data providers fail with this kind of depens_on configuration as those get evaluated to "early" - as far as I understand.

I also used depends_on in data blocks to get evaluated after databricks workspace is created.
But you have to use provider = ... in every block also to make It work.

**Comment by tbugfinder** [comment]
Finally single step apply finished successfully unsing depends_on within resources but not within the data source. In addition I haven't defined `provider=` within each resource.

**Comment by trapacampo** [comment]
> Finally single step apply finished successfully unsing depends_on within resources but not within the data source. In addition I haven't defined `provider=` within each resource.

Nice! :)
In my case I had to define provider in each block to make it work. If not, the "cannot read cluster: cannot configure azure-client-secret auth: cannot get workspace: please set azure_workspace_resource_id..." message showed up :S

**Comment by mcastorina08** [comment]
I've been dealing with this as well. I lowered the Azurerm version and eliminated the cluster error, but now i'm getting an Error building ARM Config: Authenticating using the Azure CLI is only support as a User (not a Service Principal). I tried the auth_type option, but that did nothing. Is this related?

**Comment by hedglenrm** [comment]
This is still causing issues for me.  I have specified a provider for all my blocks.  This was working about 3 weeks ago fine. 


`provider "databricks" {
  # Configuration options
  alias = "dbx"
  host                        = module.azure-databricks-workspace.workspace_url
  azure_workspace_resource_id = module.azure-databricks-workspace.id
  azure_use_msi               = true

}
data "databricks_group" "admins" {
  display_name = "admins"
  provider     = databricks.dbx
  depends_on = [
    module.azure-databricks-workspace
  ]
}`

**Comment by JCZuurmond** [comment]
TL;DR : Break down your changes in the smallest steps and **run an apply between each of the changes**

👋 Hi all, 

I was running into the same issue; I ran into this issue already almost a year ago; mitigated the problem by pinning the azurerm version; it became time to solve this issue; and it is solved! 

Herewith is an action plan for getting yourself out of this situation:

# Key takeaway
Run an apply in between each (small) change that you make. (See below, on why I think this solves the issue).

# Action plan
I pinned the azurerm version to `v2.65.0`, when upgrading I ran into the issue mentioned above. After a search, I noticed that I **could** upgrade to `v2.81.0` but **could not** to any higher version (tested for `v2.82.0`, `v2.84.0`, `v2.91.0`, `v2.92.0` and `v2.93.0`). 

Also, I did not have the `host` field set on the databricks provider. I think this is a new field; I did not need to set it before.

My action plan was:
1. Upgrade azurerm version to `v2.81.0`
2. **Run a Terraform apply**
3. Make sure the Databricks authentication is up-to-date with the [docs](https://registry.terraform.io/providers/databricks/databricks/latest/docs#special-configurations-for-azure). In my case, I needed to add the `host` field.
4. **Run a Terraform apply**
5. Upgrade azurerm to the latest version (and resolve the breaking changes of the major version upgrade)

# Hypothesis
I do not exactly know what is happening, but I think the problem originates somewhere in how Azure authenticates and connects to your (existing) Databricks workspace.

Looking at the [azurerm `v2.82.0` release notes](https://github.com/hashicorp/terraform-provider-azurerm/releases/tag/v2.82.0), they bump the version of [`github.com/Azure/azure-sdk-for-go`](https://github.com/hashicorp/terraform-provider-azurerm/pull/13613). Unfortunately, the [release notes of the SDK in go](https://github.com/Azure/azure-sdk-for-go/releases/tag/v58.0.0) do not tell us much.

However, I noticed the authentication problem occurred already when I did an `terraform plan`, more specifically when the Databricks resources were refreshed. I think that the login to Databricks changed when the Azure SDK got upgraded: it required the Databricks `host` to be stated. 

When I upgraded both the azurerm version to version `v2.82.0` (or later) AND added the `host` to the databricks provider **without running a Terraform apply in-between**, I think the Databricks `host` is not present when running the refresh while it should.

I hope this helps! Cheers!
